war/src/main/js/add-item.js

```diff
@@ -0,0 +1,386 @@
+          'top':($jenkTools.height() - 5 )+'px'});
+        $categories.css({'margin-top':$tabs.outerHeight()+'px'});
+      }
+      else{
+        $tabs.add($categories).removeAttr('style');
+      }
+    }
+
+    //////////////////////////
+    // helper functions...
+
+    function checkFormReady(){
+      //make sure everyone has changed and gotten attached...
+      setTimeout(function(){
+        var $name = $form.removeClass('no-select').find('input[name="name"]').removeClass('no-val');
+        
+        function checkItems(){
+          var selected = $form.find('input[type="radio"]:checked').length > 0;
+          var named = $.trim($name.val()).length > 0;
+          return {selected:selected,named:named};
+        }
+        if(checkItems().selected && checkItems().named){
+          $($subBtn.next('.cover')).remove();
+          $subBtn.removeClass('yui-button-disabled').find('button').removeAttr('disabled');
+        }
+        else{
+          $subBtn.addClass('yui-button-disabled').find('button').attr('disabled','disabled');
+          $('<div class="cover" />').insertAfter($subBtn)
+            .click(function(){
+              var missingText = [];
+              if(!checkItems().named){
+                $name.addClass('no-val').focus();
+                missingText.push('item name');
+              }
+              if(!checkItems().selected){
+                $form.addClass('no-select');
+                missingText.push('item type selection');
+              }
+              alert('Required selections are missing: '+ (missingText.join(', ')));
+            });
+        }
+      },10);
+    }
+    
+    function addCopyOption(data){
+      var $copy = $('#copy').closest('tr');
+      if($copy.length === 0) {return data;} // exit if copy should not be added to page. Jelly page holds that logic.
+      var copyTitle = $copy.find('label').text();
+      var copyDom = $copy.next().find('.setting-main').html();
+      var copy = {
+          name:'Copy',
+          id:'copy',
+          minToShow:0,
+          items:[
+            {
+              class:"copy",
+              description:copyDom,
+              displayName:copyTitle,
+              iconFilePathPattern:'images/48x48/copy.png'
+            }
+          ]
+      };
+      var newData = [];
+
+      $.each(data,function(i,elem){
+        if(elem.id !== "category-id-copy")
+          { newData.push(elem); }
+      });
+      
+      newData.push(copy);
+
+      return newData;
+    }
+    function checkForLink(desc){
+      if(desc.indexOf('&lt;a href="') === -1) return false;
+      var newDesc = desc.replace(/\&lt;/g,'<').replace(/\&gt;/g,'>');
+      return newDesc;
+    }
+    function sortItemsByOrder(itemTypes) {
+      function sortByOrder(a, b) {
+        var aOrder = a.weight;
+        var bOrder = b.weight;
+        return ( (aOrder < bOrder) ? -1 : ( (aOrder === bOrder) ? 0 : 1));
+      }
+      return itemTypes.sort(sortByOrder);
```

Comment:
And it is unclear to me whether we are sorting _categories_ here, as the naming in the calling code seems to suggest, or _item types_, as the naming in the function seems to suggest. Whichever it is, the server side defines `weight` for both of these, but the client side seems to be sorting only one of them. So something is fishy.